Introduce PTQ FastBC

nikita-malininn opened the pull request:
### Changes

<!--- What was changed (briefly), how to reproduce (if applicable), what the reviewers should focus on -->

### Reason for changes

<!--- Why should the change be applied -->

### Related tickets

<!--- Post the numerical ID of the ticket, if available -->

### Tests

<!--- How was the correctness of changes tested and whether new tests were added -->

nikita-malininn committed:
Introduce PTQ FastBC

nikita-malininn committed:
Rollback metatypes

nikita-malininn committed:
Fixes & removes

nikita-malininn committed:
Rollback dataset len

andrey-churkin commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,195 @@
+        transformation_commands = []
+        nncf_graph = NNCFGraphFactory.create(model)
+
+        biased_nodes = self.get_nodes_to_correct(model)
```
```suggestion
        biased_nodes = nncf_graph.get_nodes_by_metatypes(LAYERS_WITH_BIAS_METATYPES)
```

I think it is better to avoid redundant methods.

nikita-malininn replied:
Done.

dongkwan-kim01 committed:
numpy version update for openvino compatibility (#1274)

### Changes

Numpy version update from `1.22.1` to `1.23.1`

### Reason for changes

Error in the openvino test in the CI server
![image](https://user-images.githubusercontent.com/49050419/191700557-db2345a4-ab3c-4fdd-881b-3d4cfc0de0b3.png)

### Related tickets

### Tests

Testing with PR in CI

Co-authored-by: Alexander Suslov <alexander.suslov@intel.com>
Co-authored-by: Andrey Churkin <andrey.churkin@intel.com>

vshampor committed:
Refactor schemata and prepare for rendering. (#1268)

Changes
Refactored the schema file for better visualization using json-schema-for-humans. Extracted config defaults into named constants so that they are aligned between the schema and the config parsing code. Improved config parameter descriptions throughout. Added a Github Action to build and deploy an HTML rendering of the schema.

Reason for changes
The schema file was already too big and pretty much unreadable. HTML rendering of the schema should help users write better configs and better understand what wrong in their own config.

Related tickets
N/A

Tests
Existing test_config_schema tests.

vshampor committed:
Fix page build action

nikita-malininn committed:
Update FastBC and add new Command

nikita-malininn committed:
Remove ONNX mentions from FastBC

nikita-malininn committed:
Minor updates

andrey-churkin commented on nncf/experimental/onnx/graph/model_transformer.py:
```diff
@@ -61,33 +64,47 @@ def _get_transformation_layout_extra_outputs(
                 output_insert_transformations.append(transformation)
+            elif isinstance(transformation, ONNXBiasCorrectionCommand):
+                bias_correction_transformations.append(transformation)
+            elif isinstance(transformation, ONNXModelExtractionCommand):
+                model_extraction_transformation = transformation
```
Is it a valid situation when we have several `ONNXModelExtractionCommand`?

nikita-malininn replied:
In general, we shouldn't have several `ONNXModelExtractionCommand`'s. But it's a good question to discuss.

andrey-churkin commented on nncf/experimental/onnx/graph/transformations/commands.py:
```diff
@@ -70,3 +71,23 @@ class ONNXOutputInsertionCommand(ONNXInsertionCommand):
         raise NotImplementedError()
+
+class ONNXBiasCorrectionCommand(TransformationCommand):
+    def __init__(self, target_point: ONNXTargetPoint, bias_value: np.array, threshold: float):
```
```suggestion
    def __init__(self, target_point: ONNXTargetPoint, bias_value: np.ndarray, threshold: float):
```

nikita-malininn replied:
Done, thanks.

andrey-churkin commented on nncf/experimental/onnx/statistics/collectors.py:
```diff
@@ -89,3 +99,14 @@ def _register_input(self, x: ONNXNNCFTensor):
     def _get_statistics(self) -> ONNXMinMaxTensorStatistic:
         return ONNXMinMaxTensorStatistic(self._min_aggregate().tensor, self._max_aggregate().tensor)
+
+class ONNXMeanStatisticCollector(MeanStatisticCollector):
```
Is it possible to create a common `MeanStatisticCollector`?

nikita-malininn replied:
It was created similarly with `TF...StatisticCollector`, `PT...StatisticCollector` and etc. Not sure that the common way is applicable here.

andrey-churkin commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+            'Gemm': -1,
+        }
+
+    def _generate_stat_collector(self, model_backend, axes) -> TensorStatisticCollectorBase:
+        return PTQMeanStatisticCollectorFactory.create(
+            model_backend,
+            reduction_shape=axes,
+            num_samples=self.number_samples)
```
This method is unused. Could you please remove it?

nikita-malininn replied:
Done, thanks.

andrey-churkin commented on nncf/experimental/post_training/factories.py:
```diff
@@ -0,0 +1,91 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+"""
+from build.lib.nncf.experimental.onnx.graph.metatypes.onnx_metatypes import LAYERS_WITH_BIAS_METATYPES
```
```suggestion
"""

from build.lib.nncf.experimental.onnx.graph.metatypes.onnx_metatypes import LAYERS_WITH_BIAS_METATYPES
```

andrey-churkin commented on nncf/experimental/post_training/factories.py:
```diff
@@ -0,0 +1,91 @@
+ limitations under the License.
+"""
+from build.lib.nncf.experimental.onnx.graph.metatypes.onnx_metatypes import LAYERS_WITH_BIAS_METATYPES
+from build.lib.nncf.experimental.onnx.graph.metatypes.onnx_metatypes import ONNX_OPERATION_METATYPES
```
Wrong import.

nikita-malininn replied:
Done.

andrey-churkin commented on nncf/experimental/post_training/factories.py:
```diff
@@ -0,0 +1,91 @@
+    def get_layers_with_bias_types(backend):
+        if backend == BackendType.ONNX:
+            layers_with_bias = LAYERS_WITH_BIAS_METATYPES
+        return layers_with_bias
```
```suggestion
        return layers_with_bias

```

AlexKoff88 commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+    def _apply(self, model: ModelType, engine: Engine,
+               statistic_points: StatisticPointsContainer) -> ModelType:
+
+        model_backend = infer_backend_from_model(model)
```
`infer_backend_from_model` -> `get_backend`

nikita-malininn replied:
Done.

AlexKoff88 commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+        transformation_layout = TransformationLayout()
+        nncf_graph = NNCFGraphFactory.create(model)
+
+        layers_with_bias_types = PTQLayerMetatypesFactory.get_layers_with_bias_types(
```
Is this really a backend specific?

nikita-malininn replied:
To be true: I'm not sure. But as I can see, this was made similarly with the other backends. Here are the ONNX meta types - https://github.com/openvinotoolkit/nncf/blob/develop/nncf/experimental/onnx/graph/metatypes/onnx_metatypes.py
And here are the PT meta types, for example - https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/graph/operator_metatypes.py

AlexKoff88 commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+            input_shape = []
+            output_fp = []
+
+            for tensor_collector in statistic_points.iter_through_algorithm_tensor_collectors_in_target_node(
```
I don't like this interface really. It is not obvious and introduces some overheard with 4 for-loops if we look into details on how it works. Consider redesigning this in the future.

nikita-malininn replied:
Good comment. Support this, in general. For the details, it's better to ask the author of this method.

nikita-malininn committed:
Apply comments

AlexKoff88 commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+                    PostTrainingAlgorithms.FastBiasCorrection):
+                input_fp.extend(tensor_collector.get_statistics().mean_values)
+                input_shape.extend(tensor_collector.get_statistics().shape)
+            for tensor_collector in statistic_points.iter_through_algorithm_tensor_collectors_in_target_node(
```
I would split the code a little bit just for future generations), like:
- get_fp_inputs()
- get_fp_outputs()
- extract_submodel()
- get_bias_shift()
- correct_bias()

nikita-malininn replied:
Done.

AlexKoff88 commented on nncf/experimental/post_training/algorithms/quantization/fast_bias_correction.py:
```diff
@@ -0,0 +1,225 @@
+                              output_fp,
+                              output_names):
+
+        engine.rt_session_options['providers'] = ['OpenVINOExecutionProvider']
```
How it is supposed to work across different backends? or this is a part we haven't touched yet.

nikita-malininn replied:
Haven't touched it yet. Maybe it's too early to introduce this. Should we remove it?

AlexKoff88 replied:
let's discuss first

AlexKoff88 commented on nncf/experimental/post_training/factories.py:
```diff
@@ -0,0 +1,91 @@
+from nncf.experimental.onnx.tensor import ONNXNNCFTensor
+
+
+class NNCFGraphFactory:
```
It looks like we need to combine graph related factories into the one single and instantiate only one object responsible for a model transformations.

AlexKoff88 commented on nncf/experimental/post_training/factories.py:
```diff
@@ -0,0 +1,91 @@
+    @staticmethod
+    def create(backend, target_point, bias_value, threshold) -> TargetPoint:
+        if backend == BackendType.ONNX:
+            bias_correction_command = ONNXBiasCorrectionCommand(
```
Do we need backend-specific commands? I presume commands can be universal and reused across backends.

nikita-malininn replied:
This is another instance that was created similarly to the existing commands. For example - https://github.com/openvinotoolkit/nncf/blob/develop/nncf/torch/graph/transformations/commands.py

nikita-malininn committed:
Apply comments

nikita-malininn committed:
Added AlgoBackend entity

AlexKoff88 commented on nncf/experimental/post_training/algorithms/fast_bias_correction/algorithm.py:
```diff
@@ -0,0 +1,223 @@
+    
+    def _set_backend_entity(self, model):
+        model_backend = get_backend(model)
+        if model_backend == BackendType.ONNX:
```
I think we should create an interface, like `AlgoBackend`, which will contain all the backend-specific features that should be implemented to support the method in the backend.
We need also to have a registry with the available backends and provide an API inside every algo that will return the supported backends, like `algo.available_backends() -> List[Backend]`

nikita-malininn replied:
In general, should we have a common `AlgoBackend` or specific interface for each algorithm and per-backend implementations nearby?

AlexKoff88 replied:
It should be per-algorithm.

nikita-malininn committed:
Update AlgoBackend entities

AlexKoff88 commented on nncf/experimental/post_training/algorithms/fast_bias_correction/algo_backend.py:
```diff
@@ -0,0 +1,70 @@
+"""
```
I would suggest the following naming convention:
- Files:
  - `backend.py`
  - `onnx_backend.py`
  - ...
- Classes:
  - `FBCBackend()`
  - `ONNXFBCBackend()`

AlexKoff88 commented on nncf/experimental/post_training/algorithms/fast_bias_correction/onnx_algo_backend.py:
```diff
@@ -0,0 +1,67 @@
+from nncf.experimental.post_training.algorithms.fast_bias_correction.algo_backend import FBCAlgoBackend
+
+
+@ALGO_BACKENDS.register()
```
The idea was to create a enum with the list of backends: `BACKENDS.ONNX, BACKENDS.OV,BACKENDS.PT, BACKENDS.TF` and use their names at registration, like:

`@ALGO_BACKENDS.register(BACKENDS.ONNX)`

AlexKoff88 replied:
Oh, you probably just need to put `BackendType.ONNX` there.

AlexKoff88 commented on nncf/experimental/post_training/algorithms/fast_bias_correction/algorithm.py:
```diff
@@ -0,0 +1,224 @@
+        }
+        self._backend_entity = None
+
+    def available_backends(self) -> List[BackendType]:
```
I would suggest making this a property (with empty default implementation) in the base Algorithm class.

alexsu52 commented on nncf/experimental/post_training/algorithms/fast_bias_correction/algo_backend.py:
```diff
@@ -0,0 +1,70 @@
+ALGO_BACKENDS = Registry('algo_backends')
+
+
+class FBCAlgoBackend(ABC):
```
Could you add type hints for class methods? Let's discuss offline the naming of class methods.

nikita-malininn committed:
Apply comments, add tests, descriptions

nikita-malininn committed:
Merge remote-tracking branch 'origin/develop' into HEAD

nikita-malininn committed:
Apply main comments

nikita-malininn committed:
Apply comments

nikita-malininn committed:
Apply comments

nikita-malininn committed:
Apply comments & change docstrings

nikita-malininn committed:
Apply comments

nikita-malininn committed:
Fix tests

nikita-malininn committed:
Fix tests

nikita-malininn committed:
Replace magnitude calculation method into the algorithm

nikita-malininn committed:
Revert

nikita-malininn committed:
Fix bias test

AlexKoff88 commented:
I am fine with merging this. The method will be enabled in a separate PR after extensive validation.

KodiaqQ commented:
> I am fine with merging this. The method will be enabled in a separate PR after extensive validation.

Sorry for the misleading with the PR's. 
The main PR into NNCF https://github.com/openvinotoolkit/nncf/pull/1282
